simulator/main.rs

```diff
@@ -67,6 +116,343 @@ enum Value {
+                        t[i] = rng.gen_range(0..=255);
+                    }
+                    LTValue(Value::Text(String::from_utf8(t).unwrap()))
+                }
+            }
+            Value::Blob(b) => todo!(),
+            _ => unreachable!(),
+        }
+    }
+}
+
+struct GTValue(Value);
+
+impl ArbitraryFrom<Vec<&Value>> for GTValue {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &Vec<&Value>) -> Self {
+        if t.is_empty() {
+            return GTValue(Value::Null);
+        }
+
+        let index = rng.gen_range(0..t.len());
+        GTValue::arbitrary_from(rng, t[index])
+    }
+}
+
+impl ArbitraryFrom<Value> for GTValue {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &Value) -> Self {
+        match t {
+            Value::Integer(i) => GTValue(Value::Integer(rng.gen_range(*i..i64::MAX))),
+            Value::Float(f) => GTValue(Value::Float(rng.gen_range(*f..1e10))),
+            Value::Text(t) => {
+                // Either lengthen the string, or make at least one character smaller and mutate the rest
+                let mut t = t.clone();
+                if rng.gen_bool(0.01) {
+                    t.push(rng.gen_range(0..=255) as u8 as char);
+                    GTValue(Value::Text(t))
+                } else {
+                    let index = rng.gen_range(0..t.len());
+                    let mut t = t.into_bytes();
+                    t[index] += 1;
+                    // Mutate the rest of the string
+                    for i in (index + 1)..t.len() {
+                        t[i] = rng.gen_range(0..=255);
+                    }
+                    GTValue(Value::Text(String::from_utf8(t).unwrap()))
+                }
+            }
+            Value::Blob(b) => todo!(),
+            _ => unreachable!(),
+        }
+    }
+}
+
+enum Predicate {
+    And(Vec<Predicate>),
+    Or(Vec<Predicate>),
+    Eq(String, Value),
+    Gt(String, Value),
+    Lt(String, Value),
+}
+
+enum Query {
+    Create { table: Table },
+    Select { table: String, predicate: Predicate },
+    Insert { table: String, values: Vec<Value> },
+    Delete { table: String, predicate: Predicate },
+}
+
+impl ArbitraryFrom<Table> for Query {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &Table) -> Self {
+        match rng.gen_range(0..=200) {
+            0 => Query::Create {
+                table: Table::arbitrary(rng),
+            },
+            1..=100 => Query::Select {
+                table: t.name.clone(),
+                predicate: Predicate::arbitrary_from(rng, t),
+            },
+            101..=200 => Query::Insert {
+                table: t.name.clone(),
+                values: t
+                    .columns
+                    .iter()
+                    .map(|c| Value::arbitrary_from(rng, &c.column_type))
+                    .collect(),
+            },
+            201..=300 => Query::Delete {
+                table: t.name.clone(),
+                predicate: Predicate::arbitrary_from(rng, t),
```

Comment:
I believe this would be nicer if it followed the workload distribution. What do you think

Reply:
I think I missed the workload distribution part when I was working through the code. I'll roll back and try to understand how that fit into the previous scheme and bring it back.